File: advent-of-code/2021/src/day3.clj
(def path "d:/dev/ps/advent-of-code/2021/input/day3.txt")

(defn merge-bit-pair [a b]
  (merge-with + a b))

(defn merge-bits [as bs]
  (map merge-bit-pair as bs))

(defn bit-freq-per-position [s]
  (->> s
       (map (fn [line] (map (fn [c] {c 1}) line)))
       (reduce merge-bits)))

(merge-bit-pair {:a 1} {:b 1 :a 2})
(merge-bits [{:a 1} {:b 1 :a 2}] [{:a 1} {:b 1 :a 2}])

(defn bin [bs]
  (read-string (str "2r" bs)))

(bin "0")
(bin "10")
(bin "101")

(defn gamma-rate [lines]
  (->> lines
       bit-freq-per-position
       (map (fn [bit] (key (apply max-key val bit))))
       (clojure.string/join)
       (bin)))

(defn eps-rate [lines]
  (->> lines
       bit-freq-per-position
       (map (fn [bit] (key (apply min-key val bit))))
       (clojure.string/join)
       (bin)))

(def input-lines
  (with-open [r (clojure.java.io/reader path)]
    (vec (line-seq r))))

; part 1
(* (gamma-rate input-lines) (eps-rate input-lines))

(defn part-2 [lines bit-criteria pos]
  (if (== (count lines) 1)
    (bin (first lines))
    (let [keep-bit (bit-criteria (nth (bit-freq-per-position lines) pos))
          filtered (filter (fn [line] (= (nth line pos) keep-bit)) lines)]
      (part-2 filtered bit-criteria (+ pos 1)))))

(defn o2-rating [lines]
  (let [most-common-bit (fn [bit] (if (<= (bit \0) (bit \1)) \1 \0))]
    (part-2 lines most-common-bit 0)))
(defn co2-rating [lines]
  (let [least-common-bit (fn [bit] (if (<= (bit \0) (bit \1)) \0 \1))]
    (part-2 lines least-common-bit 0)))

; part 2
(* (o2-rating input-lines) (co2-rating input-lines))

File: advent-of-code/2021/src/day3.kt
import java.io.File

fun mergeBitPair(a: Map<Char, Int>, b: Map<Char, Int>): Map<Char, Int> {
    val merged = a.toMutableMap()
    b.forEach { (k, v) -> merged.compute(k) { _, v0 -> (v0 ?: 0) + v } }
    return merged
}

fun bitFreqPerPosition(lines: List<String>) =
    lines.map { line -> line.map { mapOf(it to 1) } }
        .reduce { a, b -> a.zip(b, ::mergeBitPair) }

fun gammaRate(lines: List<String>) =
    bitFreqPerPosition(lines)
        .mapNotNull { bit -> bit.maxByOrNull { it.value }?.key }
        .joinToString("")
        .toInt(2)

fun epsRate(lines: List<String>) =
    bitFreqPerPosition(lines)
        .mapNotNull { bit -> bit.minByOrNull { it.value }?.key }
        .joinToString("")
        .toInt(2)

fun part2(lines: List<String>, bitCriteria: (Map<Char, Int>) -> Char, pos: Int): List<String> =
    if (lines.size == 1) {
        lines
    } else {
        val keepBit = bitCriteria(bitFreqPerPosition(lines)[pos])
        val filtered = lines.filter { it[pos] == keepBit }
        part2(filtered, bitCriteria, pos + 1)
    }

fun part2(lines: List<String>, bitCriteria: (Map<Char, Int>) -> Char): Int =
    part2(lines, bitCriteria, 0).single().toInt(2)

fun o2Rating(lines: List<String>) =
    part2(lines) {
        if ((it['0'] ?: 0) <= (it['1'] ?: 0)) '1' else '0'
    }

fun co2Rating(lines: List<String>) =
    part2(lines) {
        if ((it['0'] ?: 0) <= (it['1'] ?: 0)) '0' else '1'
    }

fun main() {
    val inputLines = File("input/day3.txt").readLines()

    // part 1
    println(gammaRate(inputLines) * epsRate(inputLines))

    // part 2
    println(o2Rating(inputLines) * co2Rating(inputLines))
}
